dot/core/process.go

```diff
@@ -0,0 +1,292 @@
+		subchain = subchain[:maxResponseSize]
+	}
+
+	responseData := []*types.BlockData{}
+
+	for _, hash := range subchain {
+
+		blockData := new(types.BlockData)
+		blockData.Hash = hash
+
+		// set defaults
+		blockData.Header = optional.NewHeader(false, nil)
+		blockData.Body = optional.NewBody(false, nil)
+		blockData.Receipt = optional.NewBytes(false, nil)
+		blockData.MessageQueue = optional.NewBytes(false, nil)
+		blockData.Justification = optional.NewBytes(false, nil)
+
+		// header
+		if (blockRequest.RequestedData & 1) == 1 {
+			retData, err := s.blockState.GetHeader(hash)
+			if err == nil && retData != nil {
+				blockData.Header = retData.AsOptional()
+			}
+		}
+		// body
+		if (blockRequest.RequestedData&2)>>1 == 1 {
+			retData, err := s.blockState.GetBlockBody(hash)
+			if err == nil && retData != nil {
+				blockData.Body = retData.AsOptional()
+			}
+		}
+		// receipt
+		if (blockRequest.RequestedData&4)>>2 == 1 {
+			retData, err := s.blockState.GetReceipt(hash)
+			if err == nil && retData != nil {
+				blockData.Receipt = optional.NewBytes(true, retData)
+			}
+		}
+		// message queue
+		if (blockRequest.RequestedData&8)>>3 == 1 {
+			retData, err := s.blockState.GetMessageQueue(hash)
+			if err == nil && retData != nil {
+				blockData.MessageQueue = optional.NewBytes(true, retData)
+			}
+		}
+		// justification
+		if (blockRequest.RequestedData&16)>>4 == 1 {
+			retData, err := s.blockState.GetJustification(hash)
+			if err == nil && retData != nil {
+				blockData.Justification = optional.NewBytes(true, retData)
+			}
+		}
+
+		responseData = append(responseData, blockData)
+	}
+
+	return &network.BlockResponseMessage{
+		ID:        blockRequest.ID,
+		BlockData: responseData,
+	}, nil
+}
+
+// BlockResponseMessage 2
+
+// ProcessBlockResponseMessage attempts to validate and add the block to the
+// chain by calling `core_execute_block`. Valid blocks are stored in the block
+// database to become part of the canonical chain.
+func (s *Service) ProcessBlockResponseMessage(msg network.Message) error {
+	log.Debug("[core] received BlockResponseMessage")
+
+	// cast Message to BlockResponseMessage in order to create header
+	blockResponseMessage, ok := msg.(*network.BlockResponseMessage)
+	if !ok {
+		return errors.New("could not cast Message to BlockResponseMessage")
+	}
+
+	// loop through block data in block response message
+	for _, bd := range blockResponseMessage.BlockData {
+		if bd.Header.Exists() {
+
+			header, err := types.NewHeader(
+				bd.Header.Value().ParentHash,
+				bd.Header.Value().Number,
+				bd.Header.Value().StateRoot,
+				bd.Header.Value().ExtrinsicsRoot,
+				bd.Header.Value().Digest,
+			)
+			if err != nil {
+				return err
+			}
+
+			// check authority
+			if s.isAuthority {
+
+				// check if block is from current epoch
+				err = s.verifyCurrentEpoch(header)
+				if err != nil {
+					log.Trace("[core] block is not from current epoch", "error", err)
+				}
```

Comment:
would this work if we are syncing? eg. the chain is at epoch 100 and we are still syncing past epochs for example

Reply:
related to https://github.com/ChainSafe/gossamer/pull/703#discussion_r398783470

Reply:
could you explain here? the comments just link to each other lol

Reply:
I was still addressing this, revised in latest commit